# WORKSHEET 0

## Name: Leeja Sagar Shrestha



## Group: L6CG7

### 3.1 About Python

Python Version Check

In [1]:
import sys

# Check Python version
if sys.version_info.major == 3 and sys.version_info.minor >= 6:
    print("Hello, World!")
    print(f"Python version: {sys.version}")
else:
    print("Please use Python 3.6 or higher.")


Hello, World!
Python version: 3.11.11 (main, Dec  4 2024, 08:55:07) [GCC 11.4.0]


3.2 Data Types in Python


Difference between Mutable and Immutable Data Types.

In [4]:
# Mutable Data Type Example (List)
my_list = [1, 2, 3]
print("Original List:", my_list)
# Modifying the list (changing contents)
my_list.append(4)
print("Modified List:", my_list)
# Immutable Data Type Example (Tuple)
my_tuple = (1, 2, 3)
print("Original Tuple:", my_tuple)

Original List: [1, 2, 3]
Modified List: [1, 2, 3, 4]
Original Tuple: (1, 2, 3)


Once a tuple is created, its elements cannot be modified, added, or removed.

3.2.1 Some Common Data Types in Python:
Data Types - Numeric

In [3]:
age = 23 #int
pi = 3.14 #float
temperature = -5.5 #float
complex_number = 2+3j #COMPLEX DATA TYPE
print("data type of variable age = ", type(age))
print("data type of variable pi = ", type(pi))
print("data type of variable temperature = ", type(temperature))
print("Data type of variable complex_number =", type(complex_number))

data type of variable age =  <class 'int'>
data type of variable pi =  <class 'float'>
data type of variable temperature =  <class 'float'>
Data type of variable complex_number = <class 'complex'>


SEQUENCE: Data Types - Sequence - String

In [5]:
name = "Alice"
greeting = 'Hello'
address = "123 Main St"
print("data type of the variable name = ",type(name))
print("data type of the variable greeting = ",type(greeting))
print("data type of the variable address = ",type(address))
# slice only one element
print("The first letter of the name is:", name[0])
print("The last letter of the name is:", name[-1])
# slice a range of elements
print("The second letter to the fourth of the name is:", name[1:4])
print("The first two letters of the name are:", name[:2])
print("Substring starting from the third letter is:", name[2:])

data type of the variable name =  <class 'str'>
data type of the variable greeting =  <class 'str'>
data type of the variable address =  <class 'str'>
The first letter of the name is: A
The last letter of the name is: e
The second letter to the fourth of the name is: lic
The first two letters of the name are: Al
Substring starting from the third letter is: ice


Data Types - Sequence - List

In [7]:
list1 = [1, 2, 3, 4]  # A list containing integers
mixed_list = [12, "Hello", True]  # A list containing different data types
# List is mutable, so we change the first element (index 0)
mixed_list[0] = False
# Output the modified list
mixed_list


[False, 'Hello', True]

Data Types - Sequence - Tuple

In [8]:
colors = ('red', 'green', 'yellow', 'blue')
print("First element:", colors[0])
print("Last two elements:", colors[2:])
print("Middle two elements:", colors[1:3])
colors[0] = 'purple'
colors # will generate an error as tuple is immutable.

First element: red
Last two elements: ('yellow', 'blue')
Middle two elements: ('green', 'yellow')


TypeError: 'tuple' object does not support item assignment

MAPPING: Data Types - Sequence - Dict

In [9]:
person = {'name':'John','age':30,'city':'Pittsburgh'}
print(f"Hello my name is {person['name']}. I am {person['age']} years old and I live at {person['city']}.")
print("All keys:", list(person.keys()))
print("All values:", list(person.values()))

Hello my name is John. I am 30 years old and I live at Pittsburgh.
All keys: ['name', 'age', 'city']
All values: ['John', 30, 'Pittsburgh']


SET: Data Types - SET

In [10]:
unique_numbers = {1,2,3,3,3,3,4,5,4,2}
print("The unique numbers is:", unique_numbers)

The unique numbers is: {1, 2, 3, 4, 5}


BOOLEAN:

In [11]:
is_student = True
has_license = False
print("data type of the variable is_student = ",type(is_student))
print("data type of the variable has_license = ",type(has_license))

data type of the variable is_student =  <class 'bool'>
data type of the variable has_license =  <class 'bool'>


## 3.3 Logical Statements and Loops

## 3.4 Logical Statements

In [12]:
num = 10
if num > 0:
 print("Positive")
elif num == 0:
 print("Zero")
else:
 print("Non-positive")


Positive


Comparision Operators

In [13]:
x = 5
y = 10
if x > 0 and y < 20:
 print("Both conditions are true")
if x > 0 or y > 20:
 print("At least one condition is true")
if not x == 0:
 print("x is not equal to 0")


Both conditions are true
At least one condition is true
x is not equal to 0


Logical Operators

In [14]:
# Define two boolean variables
a = True
b = False
# Using the ’and’ operator
if a and b:
 print("Both a and b are True")
else:
 print("Either a or b is False") # This will be printed
# Using the ’or’ operator
if a or b:
 print("At least one of a or b is True") # This will be printed
else:
 print("Both a and b are False")
# Using the ’not’ operator
if not a:
 print("a is False") # This will not be printed
else:
 print("a is True") # This will be printed


Either a or b is False
At least one of a or b is True
a is True


### 3.4.1 Loops

In [15]:
# For loop:
fruits = ["apple", "banana", "cherry"]
for fruit in fruits:
    print(fruit)

# While loop:
count = 0
while count < 5:
    print("Count is:", count)
    count += 1  # Ensure count is incremented inside the loop

# Break
fruits = ["apple", "banana", "orange"]
print("loop 1")
for fruit in fruits:
    print(fruit)
    if fruit == "apple":
        break  # Stop the loop when fruit is "apple"

# Continue
print("loop 2")
for fruit in fruits:
    if fruit == "apple":
        continue  # Skip the iteration when fruit is "apple"
    else:
        print(fruit)


apple
banana
cherry
Count is: 0
Count is: 1
Count is: 2
Count is: 3
Count is: 4
loop 1
apple
loop 2
banana
orange


## 3.6 Functions

In [16]:
def add_binary(a, b):
    '''
    Returns the sum of two decimal numbers in binary digits.

    Parameters:
    a (int): A decimal integer
    b (int): Another decimal integer

    Returns:
    binary_sum (str): Binary string of the sum of a and b
    '''
    binary_sum = bin(a + b)[2:]  # bin() returns a binary string starting with '0b', so we remove '0b'
    return binary_sum
    # Example 1
print(add_binary(5, 3))  # Output: '1000' bin(8) gives '0b1000', and slicing it with [2:] gives '1000'.

# Example 2
print(add_binary(10, 2))  # Output: '1100'

# Example 3
print(add_binary(0, 0))   # Output: '0'


1000
1100
0


Built In Functions

In [17]:
# Printing "Hello, World!"
print("Hello, World!")  # Output: Hello, World!

# Finding the length of the string "Hello"
length = len("Hello")
print(length)  # Output: 5

# Checking the data type of the number 42
data_type = type(42)
print(data_type)  # Output: <class 'int'>

# Summing up the numbers in the list [1, 2, 3]
total = sum([1, 2, 3])
print(total)  # Output: 6


Hello, World!
5
<class 'int'>
6


Built in methods

In [18]:
# str.upper(): Converts all characters in a string to uppercase.
message = "hello".upper()
print(message)  # Output: "HELLO"

# list.append(): Adds an element to the end of a list.
fruits = ["apple", "banana"]
fruits.append("cherry")
print(fruits)  # Output: ["apple", "banana", "cherry"]

# dict.get(): Returns the value associated with a key in a dictionary.
info = {"name": "Alice", "age": 25}
age = info.get("age")
print(age)  # Output: 25


HELLO
['apple', 'banana', 'cherry']
25


User Defined Functions

In [19]:
def temperature_converter():
    """
    Converts temperature between Celsius and Fahrenheit.
    This function prompts the user to specify the conversion type (Celsius to Fahrenheit or Fahrenheit to Celsius)
    and then asks for the temperature value. It calculates and returns the converted temperature.

    Returns:
    float: The converted temperature value.
    """
    print("Choose conversion type:")
    print("1. Celsius to Fahrenheit")
    print("2. Fahrenheit to Celsius")

    # Get conversion choice from user
    choice = input("Enter 1 or 2: ")

    if choice == "1":
        # Celsius to Fahrenheit conversion
        try:
            celsius = float(input("Enter temperature in Celsius: "))
            fahrenheit = (celsius * 9/5) + 32
            print(f"{celsius}C is equal to {fahrenheit}F")
            return fahrenheit
        except ValueError:
            print("Invalid input. Please enter a valid number for Celsius.")
            return None

    elif choice == "2":
        # Fahrenheit to Celsius conversion
        try:
            fahrenheit = float(input("Enter temperature in Fahrenheit: "))
            celsius = (fahrenheit - 32) * 5/9
            print(f"{fahrenheit}F is equal to {celsius}C")
            return celsius
        except ValueError:
            print("Invalid input. Please enter a valid number for Fahrenheit.")
            return None

    else:
        print("Invalid choice. Please enter 1 or 2.")
        return None

# Call the function
temperature_converter()


Choose conversion type:
1. Celsius to Fahrenheit
2. Fahrenheit to Celsius
Enter 1 or 2: 1
Enter temperature in Celsius: 26
26.0C is equal to 78.8F


78.8

3.7 Global and Local Variable

GLOBAL

In [22]:
x = 10
def print_global():
 print(x)
print_global()


10


LOCAL

In [28]:
def print_local():
    y = 5
    print(y)

print_local()

5


In [24]:
x = 10 # Global variable
def modify_global():
 global x # Declaring x as global
x = 20 # Modifying global x
modify_global()
print(x) # Output: 20

20


## 3.8 Exception and Error Handling

In [30]:
try:
    num = int(input("Enter a number: "))  # Attempt to convert input to an integer
    print("Result:", 10 / num)  # Attempt to divide 10 by the input number
except ZeroDivisionError:
    print("Cannot divide by zero.")  # Handles division by zero error
except ValueError:
    print("Invalid input. Please enter a number.")  # Handles invalid input error


Enter a number: 0
Cannot divide by zero.


In [31]:
try:
    file = open("example.txt", "r")  # Try to open the file in read mode
    data = file.read()  # Read the content of the file
except FileNotFoundError:
    print("File not found.")  # Error message if the file is not found
else:
    # If no exceptions, print file data
    print(data)
finally:
    try:
        file.close()  # Ensure the file is closed after reading or error
    except NameError:
        pass  # If file was not opened (due to exception), do nothing


File not found.


In [32]:
def check_age(age):
    if age < 0:
        raise ValueError("Age cannot be negative.")  # Raise error if age is negative
    else:
        print("Valid age")  # Print if the age is valid (non-negative)

try:
    check_age(-1)  # Test with a negative age value
except ValueError as e:
    print(e)  # Print the error message if ValueError is raised


Age cannot be negative.


# TO DO TASK

## 4.1 Exercise on Functions:

Task - 1:
Create a Python program that converts between different units of measurement. • The program should:
1. Prompt the user to choose the type of conversion (e.g., length, weight, volume). 2. Ask the user to input the value to be converted.
3. Perform the conversion and display the result.
4. Handle potential errors, such as invalid input or unsupported conversion types. • Requirements:
1. Functions: Define at least one function to perform the conversion.
2. Error Handling: Use try-except blocks to handle invalid input (e.g., non-numeric values). 3. User Input: Prompt the user to select the conversion type and input the value.
4. Docstrings: Include a docstring in your function to describe its purpose, parameters, and return value.
• Conversion Options:
1. Length:
– Convert meters (m) to feet (ft).
– Convert feet (ft) to meters (m).
2. Weight:
– Convert kilograms (kg) to pounds (lbs).
– Convert pounds (lbs) to kilograms (kg).
3. Volume:
– Convert liters (L) to gallons (gal).
– Convert gallons (gal) to liters (L).


In [33]:
def convert_units(value, conversion_type):

    if conversion_type == 'm_to_ft':
        return value * 3.28084
    elif conversion_type == 'ft_to_m':
        return value / 3.28084
    elif conversion_type == 'kg_to_lbs':
        return value * 2.20462
    elif conversion_type == 'lbs_to_kg':
        return value / 2.20462
    elif conversion_type == 'L_to_gal':
        return value * 0.264172
    elif conversion_type == 'gal_to_L':
        return value / 0.264172
    else:
        raise ValueError("Invalid conversion type.")


def main():
    print("Unit Conversion Program")
    print("Choose the type of conversion:")
    print("1. Length")
    print("2. Weight")
    print("3. Volume")

    try:
        choice = int(input("Enter your choice (1, 2, or 3): "))
        if choice not in [1, 2, 3]:
            print("Invalid choice. Please select 1, 2, or 3.")
            return
        value = float(input("Enter the value to convert: "))
        if choice == 1:
            print("Length Conversions:")
            print("1. Meters to Feet")
            print("2. Feet to Meters")
            sub_choice = int(input("Enter your choice (1 or 2): "))
            if sub_choice == 1:
                result = convert_units(value, 'm_to_ft')
                print(f"{value} meters = {result:.2f} feet")
            elif sub_choice == 2:
                result = convert_units(value, 'ft_to_m')
                print(f"{value} feet = {result:.2f} meters")
            else:
                print("Invalid choice.")
        elif choice == 2:
            print("Weight Conversions:")
            print("1. Kilograms to Pounds")
            print("2. Pounds to Kilograms")
            sub_choice = int(input("Enter your choice (1 or 2): "))
            if sub_choice == 1:
                result = convert_units(value, 'kg_to_lbs')
                print(f"{value} kilograms = {result:.2f} pounds")
            elif sub_choice == 2:
                result = convert_units(value, 'lbs_to_kg')
                print(f"{value} pounds = {result:.2f} kilograms")
            else:
                print("Invalid choice.")
        elif choice == 3:
            print("Volume Conversions:")
            print("1. Liters to Gallons")
            print("2. Gallons to Liters")
            sub_choice = int(input("Enter your choice (1 or 2): "))
            if sub_choice == 1:
                result = convert_units(value, 'L_to_gal')
                print(f"{value} liters = {result:.2f} gallons")
            elif sub_choice == 2:
                result = convert_units(value, 'gal_to_L')
                print(f"{value} gallons = {result:.2f} liters")
            else:
                print("Invalid choice.")
    except ValueError as e:
        print(f"Error: {e}. Please enter a valid number.")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")

if __name__ == "__main__":
    main()

Unit Conversion Program
Choose the type of conversion:
1. Length
2. Weight
3. Volume
Enter your choice (1, 2, or 3): 3
Enter the value to convert: 1
Volume Conversions:
1. Liters to Gallons
2. Gallons to Liters
Enter your choice (1 or 2): 1
1.0 liters = 0.26 gallons


Task - 2:
Create a Python program that performs various mathematical operations on a list of numbers. • The Program should:
1. Prompt the user to choose an operation (e.g., find the sum, average, maximum, or minimum of the numbers).
2. Ask the user to input a list of numbers (separated by spaces).
18

3. Perform the selected operation and display the result.
4. Handle potential errors, such as invalid input or empty lists.
• Requirements:
1. Functions: Define at least one function for each operation (sum, average, maximum, mini mum).
2. Error Handling: Use try-except blocks to handle invalid input (e.g., non-numeric values or empty lists).
3. User Input: Prompt the user to select the operation and input the list of numbers.
4. Docstrings: Include a docstring in each function to describe its purpose, parameters, and return value.


In [35]:
def calculate_sum(numbers):

    return sum(numbers)

def calculate_average(numbers):

    if len(numbers) == 0:
        raise ValueError("Cannot calculate average of an empty list.")
    return sum(numbers) / len(numbers)

def find_maximum(numbers):

    if len(numbers) == 0:
        raise ValueError("Cannot find maximum of an empty list.")
    return max(numbers)

def find_minimum(numbers):

    if len(numbers) == 0:
        raise ValueError("Cannot find minimum of an empty list.")
    return min(numbers)

def main():
    print("Mathematical Operations Program")
    print("Choose an operation:")
    print("1. Sum")
    print("2. Average")
    print("3. Maximum")
    print("4. Minimum")

    try:
        choice = int(input("Enter your choice (1, 2, 3, or 4): "))
        if choice not in [1, 2, 3, 4]:
            print("Invalid choice. Please select 1, 2, 3, or 4.")
            return
        input_numbers = input("Enter a list of numbers separated by spaces: ")
        numbers = [float(num) for num in input_numbers.split()]

        if len(numbers) == 0:
            raise ValueError("Empty list provided. Please enter at least one number.")
        if choice == 1:
            result = calculate_sum(numbers)
            print(f"Sum of the numbers: {result}")
        elif choice == 2:
            result = calculate_average(numbers)
            print(f"Average of the numbers: {result:.2f}")
        elif choice == 3:
            result = find_maximum(numbers)
            print(f"Maximum of the numbers: {result}")
        elif choice == 4:
            result = find_minimum(numbers)
            print(f"Minimum of the numbers: {result}")
    except ValueError as e:
        print(f"Error: {e}")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
if __name__ == "__main__":
    main()

Mathematical Operations Program
Choose an operation:
1. Sum
2. Average
3. Maximum
4. Minimum
Enter your choice (1, 2, 3, or 4): 1
Enter a list of numbers separated by spaces: 1 2
Sum of the numbers: 3.0


4.2 EXERCISE ON LIST MANIPULATION

Extract Every Other Element

Write a Python function that extracts every other element from a list, starting from the first element. • Requirements:
– Define a function extract every other(lst) that takes a list lst as input and returns a new list containing every other element from the original list.
– Example: For the input [1, 2, 3, 4, 5, 6], the output should be [1, 3, 5].


In [36]:
def extract_every_other(lst):

    return lst[::2] #Start at index 0 (the first element).Step by 2 (skip every second element).

input_list = [1, 2, 3, 4, 5, 6]
result = extract_every_other(input_list)
print(result)  # Output: [1, 3, 5]

[1, 3, 5]


Slice a Sublist

Write a Python function that returns a sublist from a given list, starting from a specified index and ending at another specified index.
• Requirements:
– Define a function get sublist(lst, start, end) that takes a list lst, a starting index start, and an ending index end as input and returns the sublist from start to end (inclusive).
– Example: For the input [1, 2, 3, 4, 5, 6] with start=2 and end=4, the output should be [3, 4, 5].


In [37]:
def get_sublist(lst, start, end):

    return lst[start:end + 1]

input_list = [1, 2, 3, 4, 5, 6]
start_index = 2
end_index = 4
result = get_sublist(input_list, start_index, end_index)
print(result)  # Output: [3, 4, 5]

[3, 4, 5]


Reverse a List Using Slicing:

Write a Python function that reverses a list using slicing.
• Requirements:
– Define a function reverse list(lst) that takes a list lst and returns a reversed list using slicing.
– Example: For the input [1, 2, 3, 4, 5], the output should be [5, 4, 3, 2, 1]. 19


In [38]:
def reverse_list(lst):

    return lst[::-1]


# Example usage:
input_list = [1, 2, 3, 4, 5]
result = reverse_list(input_list)
print(result)  # Output: [5, 4, 3, 2, 1]

[5, 4, 3, 2, 1]


Remove the First and Last Elements: Write a Python function that removes the first and last elements of a list and returns the resulting sublist.
• Requirements:
– Define a function remove first last(lst) that takes a list lst and returns a sublist without the first and last elements using slicing.
– Example: For the input [1, 2, 3, 4, 5], the output should be [2, 3, 4].



In [39]:
def remove_first_last(lst):

    return lst[1:-1]

input_list = [1, 2, 3, 4, 5]
result = remove_first_last(input_list)
print(result)  # Output: [2, 3, 4]

[2, 3, 4]


Get the First n Elements:
Write a Python function that extracts the first n elements from a list.
• Requirements:
– Define a function get first n(lst, n) that takes a list lst and an integer n as input and returns the first n elements of the list using slicing.
– Example: For the input [1, 2, 3, 4, 5] with n=3, the output should be [1, 2, 3].


In [40]:
def get_first_n(lst, n):

    return lst[:n]

input_list = [1, 2, 3, 4, 5]
n = 3
result = get_first_n(input_list, n)
print(result)  # Output: [1, 2, 3]

[1, 2, 3]


Extract Elements from the End:
Write a Python function that extracts the last n elements of a list using slicing. • Requirements:
– Define a function get last n(lst, n) that takes a list lst and an integer n as input and returns the last n elements of the list.
– Example: For the input [1, 2, 3, 4, 5] with n=2, the output should be [4, 5].


In [41]:
def get_last_n(lst, n):

    return lst[-n:]

input_list = [1, 2, 3, 4, 5]
n = 2
result = get_last_n(input_list, n)
print(result)  # Output: [4, 5]

[4, 5]


Extract Elements in Reverse Order:
Write a Python function that extracts a list of elements in reverse order starting from the second-to-last element and skipping one element in between.
• Requirements:
– Define a function reverse skip(lst) that takes a list lst and returns a new list containing every second element starting from the second-to-last, moving backward.
– Example: For the input [1, 2, 3, 4, 5, 6], the output should be [5, 3, 1]. 20


In [42]:
def reverse_skip(lst):

    return lst[-2::-2]

input_list = [1, 2, 3, 4, 5, 6]
result = reverse_skip(input_list)
print(result)  # Output: [5, 3, 1]

[5, 3, 1]


### 4.3 Exercise on Nested List:

1. Flatten a Nested List:
Write a Python function that takes a nested list and flattens it into a single list, where all the elements are in a single dimension.
• Requirements:
– Define a function flatten(lst) that takes a nested list lst and returns a flattened version of the list.
– Example: For the input [[1, 2], [3, 4], [5]], the output should be [1, 2, 3, 4, 5].


In [43]:
def flatten(lst):

    flattened = []
    for item in lst:
        if isinstance(item, list):
            flattened.extend(flatten(item))  # Recursively flatten nested lists
        else:
            flattened.append(item)  # Add non-list items directly
    return flattened

input_list = [[1, 2], [3, 4], [5]]
result = flatten(input_list)
print(result)  # Output: [1, 2, 3, 4, 5]

[1, 2, 3, 4, 5]


2. Accessing Nested List Elements:
Write a Python function that extracts a specific element from a nested list given its indices. • Requirements:
– Define a function access nested element(lst, indices) that takes a nested list lst and a list of indices indices, and returns the element at that position.
– Example: For the input lst = [[1, 2, 3], [4, 5, 6], [7, 8, 9]] with indices = [1, 2], the output should be 6.


In [44]:
def access_nested_element(lst, indices):
    current = lst
    for index in indices:
        if not isinstance(current, list):
            raise TypeError("Invalid nested structure for the given indices.")
        if index < 0 or index >= len(current):
            raise IndexError("Index out of bounds.")
        current = current[index]
    return current

input_list = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
indices = [1, 2]
result = access_nested_element(input_list, indices)
print(result)  # Output: 6

6


3. Sum of All Elements in a Nested List:
Write a Python function that calculates the sum of all the numbers in a nested list (regardless of depth). • Requirements:
– Define a function sum nested(lst) that takes a nested list lst and returns the sum of all the elements.
– Example: For the input [[1, 2], [3, [4, 5]], 6], the output should be 21.


In [45]:
def sum_nested(lst):

    total = 0
    for item in lst:
        if isinstance(item, list):
            total += sum_nested(item)  # Recursively sum nested lists
        else:
            total += item  # Add non-list items directly
    return total

input_list = [[1, 2], [3, [4, 5]], 6]
result = sum_nested(input_list)
print(result)  # Output: 21

21


4. Remove Specific Element from a Nested List:
Write a Python function that removes all occurrences of a specific element from a nested list. • Requirements:
– Define a function remove element(lst, elem) that removes elem from lst and returns the modified list.
– Example: For the input lst = [[1, 2], [3, 2], [4, 5]] and elem = 2, the output should be [[1], [3], [4, 5]].


In [46]:
def remove_element(lst, elem):

    result = []
    for item in lst:
        if isinstance(item, list):
            # Recursively remove the element from nested lists
            result.append(remove_element(item, elem))
        elif item != elem:
            # Add the item to the result if it is not the element to remove
            result.append(item)
    return result

input_list = [[1, 2], [3, 2], [4, 5]]
element_to_remove = 2
result = remove_element(input_list, element_to_remove)
print(result)  # Output: [[1], [3], [4, 5]]

[[1], [3], [4, 5]]


5. Find the Maximum Element in a Nested List:
Write a Python function that finds the maximum element in a nested list (regardless of depth). • Requirements:
– Define a function find max(lst) that takes a nested list lst and returns the maximum element.
– Example: For the input [[1, 2], [3, [4, 5]], 6], the output should be 6. 21


In [47]:
def find_max(lst):
    max_value = float('-inf')  # Initialize with negative infinity
    for item in lst:
        if isinstance(item, list):
            # Recursively find the maximum in nested lists
            nested_max = find_max(item)
            if nested_max > max_value:
                max_value = nested_max
        else:
            # Compare non-list items directly
            if item > max_value:
                max_value = item
    return max_value

input_list = [[1, 2], [3, [4, 5]], 6]
result = find_max(input_list)
print(result)  # Output: 6

6


6. Count Occurrences of an Element in a Nested List:
Write a Python function that counts how many times a specific element appears in a nested list. • Requirements:
– Define a function count occurrences(lst, elem) that counts the occurrences of elem in the nested list lst.
– Example: For the input lst = [[1, 2], [2, 3], [2, 4]] and elem = 2, the output should be 3.



In [48]:
def count_occurrences(lst, elem):
    count = 0
    for item in lst:
        if isinstance(item, list):
            # Recursively count occurrences in nested lists
            count += count_occurrences(item, elem)
        elif item == elem:
            # Increment count if the item matches the element
            count += 1
    return count

input_list = [[1, 2], [2, 3], [2, 4]]
element_to_count = 2
result = count_occurrences(input_list, element_to_count)
print(result)  # Output: 3

3


7. Flatten a List of Lists of Lists:
Write a Python function that flattens a list of lists of lists into a single list, regardless of the depth. • Requirements:
– Define a function deep flatten(lst) that takes a deeply nested list lst and returns a single flattened list.
– Example: For the input [[[1, 2], [3, 4]], [[5, 6], [7, 8]]], the output should be [1, 2, 3, 4, 5, 6, 7, 8].


In [49]:
def deep_flatten(lst):

    flattened = []
    for item in lst:
        if isinstance(item, list):
            # Recursively flatten nested lists
            flattened.extend(deep_flatten(item))
        else:
            # Add non-list items directly
            flattened.append(item)
    return flattened

input_list = [[[1, 2], [3, 4]], [[5, 6], [7, 8]]]
result = deep_flatten(input_list)
print(result)  # Output: [1, 2, 3, 4, 5, 6, 7, 8]

[1, 2, 3, 4, 5, 6, 7, 8]


8. Nested List Average:
Write a Python function that calculates the average of all elements in a nested list. • Requirements:
– Define a function average nested(lst) that takes a nested list lst and returns the average of all the elements.
– Example: For the input [[1, 2], [3, 4], [5, 6]], the output should be 3.5. 22


In [50]:
def average_nested(lst):
    def flatten(lst):
        """Helper function to flatten the nested list."""
        flattened = []
        for item in lst:
            if isinstance(item, list):
                flattened.extend(flatten(item))
            else:
                flattened.append(item)
        return flattened

    # Flatten the nested list
    flattened_list = flatten(lst)

    # Calculate the average
    if not flattened_list:
        return 0  # Return 0 if the list is empty to avoid division by zero
    return sum(flattened_list) / len(flattened_list)
input_list = [[1, 2], [3, 4], [5, 6]]
result = average_nested(input_list)
print(result)  # Output: 3.5

3.5


# 5 NUMPY

# 6 Getting started with NumPy

In [51]:
import numpy as np
# initialize an all zero array with size 2 X 3:
zeros_arr = np.zeros((2,3))
print(" A zeros array is \n", zeros_arr, "with dimensions", zeros_arr.shape, "\n")

 A zeros array is 
 [[0. 0. 0.]
 [0. 0. 0.]] with dimensions (2, 3) 



Creating and Displaying Different Type of Arrays

In [52]:
import numpy as np
# Create and display zero, one, and n-dimensional arrays
zero_dim_array = np.array(5)
one_dim_array = np.array([1, 2, 3])
n_dim_array = np.array([[1, 2], [3, 4]])
for arr in [zero_dim_array, one_dim_array, n_dim_array]:
 print(f"Array:\n{arr}\nDimension: {arr.ndim}\nData type: {arr.dtype}\n")


Array:
5
Dimension: 0
Data type: int64

Array:
[1 2 3]
Dimension: 1
Data type: int64

Array:
[[1 2]
 [3 4]]
Dimension: 2
Data type: int64



1. .ndim Number of Dimensions

In [53]:
# For a 1-D array:
import numpy as np
arr_1d = np.array([1, 2, 3, 4, 5])
print(arr_1d.ndim)  # Output: 1

# For a 2-D array:
arr_2d = np.array([[1, 2, 3],
                   [4, 5, 6]])
print(arr_2d.ndim)  # Output: 2

# For a 3-D array:
arr_3d = np.array([[[1, 2, 3],
                   [4, 5, 6]],
                  [[7, 8, 9],
                   [10, 11, 12]]])
print(arr_3d.ndim)  # Output: 3

1
2
3


2 .shape Shape of the Array

In [54]:
print(arr_1d.shape) # Output: (5,)
print(arr_2d.shape) # Output: (2, 3)
print(arr_3d.shape) # Output: (2, 2, 3)

(5,)
(2, 3)
(2, 2, 3)


3 .size Total Number of Elements

In [55]:
print(arr_1d.size) # Output: 5
print(arr_2d.size) # Output: 6
print(arr_3d.size) # Output: 12


5
6
12


4. .dtype Data Type of Elements

In [56]:
print(arr_1d.dtype) # Output: int64 (or int32)
arr_float = np.array([1.1, 2.2, 3.3])
print(arr_float.dtype) # Output: float64

int64
float64


5 .itemsize Size of Each Element in Bytes

In [57]:
print(arr_1d.itemsize) # Output: 8 (for int64)

8


6. .nbytes Total Memory Consumption

In [58]:
print(arr_1d.nbytes) # Output: 40 (5 elements \times 8 bytes each for int64)

40


7. .T Transpose of the Array

In [59]:
print(arr_2d.T)
# Output:
# [[1 4]
# [2 5]
# [3 6]]


[[1 4]
 [2 5]
 [3 6]]


8. .flat Flattening the Array

In [60]:
for item in arr_2d.flat:
 print(item, end=" ") # Output: 1 2 3 4 5 6

1 2 3 4 5 6 

9. .real and .image

In [61]:
import numpy as np

arr_complex = np.array([1+2j, 3+4j])
print(arr_complex.real)  # Output: [1. 3.]
print(arr_complex.imag)  # Output: [2. 4.]

[1. 3.]
[2. 4.]


10. .astype(dtype) Convert Data Type

In [62]:
arr_float = arr_2d.astype(float)
print(arr_float)
# Output:
# [[1. 2. 3.]
# [4. 5. 6.]]


[[1. 2. 3.]
 [4. 5. 6.]]


6.1.2 Common Errors with Key Attributes

In [63]:
import numpy as np
# Original list
lst = [[1, 2], [3, 4]]
arr = np.array(lst)
print(arr.shape)  # Output: (2, 2)

(2, 2)


2. IndexError: tuple index out of range

In [64]:
import numpy as np

# Create a 1D array
arr = np.array([1, 2, 3])
if len(arr.shape) > 1:
    print(arr.shape[1])  # Access only if it's valid
else:
    print("Array is 1D and has no second dimension.")

Array is 1D and has no second dimension.


3. TypeError: ‘tuple’

In [65]:
import numpy as np

# Create a 2D array
arr = np.array([[1, 2, 3], [4, 5, 6]])
print(arr.shape)  # Output: (2, 3)

(2, 3)


4. ValueError

In [66]:
import numpy as np

a = np.array([[1, 2], [3, 4]])  # Shape: (2, 2)
b = np.array([1, 2, 3])          # Shape: (3,)

print(a + b)  # This will raise a ValueError

ValueError: operands could not be broadcast together with shapes (2,2) (3,) 

Fixing error

In [67]:
import numpy as np

a = np.array([[1, 2], [3, 4]])  # Shape: (2, 2)
b = np.array([1, 2])             # Shape: (2,)
b = b.reshape(2, 1)              # Reshape b to (2, 1)

print(a + b)

[[2 3]
 [5 6]]


5. MemoryError:

In [ ]:
import numpy as np
large_array = np.ones((100000, 100000)) # May cause MemoryError

In [1]:
import numpy as np
large_array = np.ones((10, 10)) # May cause MemoryError

### Reshaping an Array

In [2]:
import numpy as np
# Create an array with shape (2, 3)
array = np.array([[1, 2, 3],
[4, 5, 6]])
# Reshape to (3, 2), keeping 6 elements
reshaped_array = array.reshape(3, 2)
print("Original Shape:", array.shape, "\nReshaped Shape:", reshaped_array.shape) # Expected Outputs:
# Original Shape: (2, 3)
# Reshaped Shape: (3, 2)


Original Shape: (2, 3) 
Reshaped Shape: (3, 2)


.reshape - ValueError.

In [3]:
array = np.array([[1, 2, 3], [4, 5, 6]])
reshaped_array = array.reshape(4, 2)
# Error: cannot reshape array of size 6 into shape (4, 2)


ValueError: cannot reshape array of size 6 into shape (4,2)

## 7 How to Create an Array {ndarray}?

In [4]:
import numpy as np
a = np.arange(1, 10)
print(a)
x = range(1, 10)
print(x) # x is an iterator
print(list(x))
# further arange examples:
x = np.arange(10.4)
print(x)
x = np.arange(0.5, 10.4, 0.8)
print(x)


[1 2 3 4 5 6 7 8 9]
range(1, 10)
[1, 2, 3, 4, 5, 6, 7, 8, 9]
[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]
[ 0.5  1.3  2.1  2.9  3.7  4.5  5.3  6.1  6.9  7.7  8.5  9.3 10.1]


linspace

In [5]:
import numpy as np
# 50 values between 1 and 10:
print(np.linspace(1, 10))
# 7 values between 1 and 10:
print(np.linspace(1, 10, 7))
# excluding the endpoint:
print(np.linspace(1, 10, 7, endpoint=False))


[ 1.          1.18367347  1.36734694  1.55102041  1.73469388  1.91836735
  2.10204082  2.28571429  2.46938776  2.65306122  2.83673469  3.02040816
  3.20408163  3.3877551   3.57142857  3.75510204  3.93877551  4.12244898
  4.30612245  4.48979592  4.67346939  4.85714286  5.04081633  5.2244898
  5.40816327  5.59183673  5.7755102   5.95918367  6.14285714  6.32653061
  6.51020408  6.69387755  6.87755102  7.06122449  7.24489796  7.42857143
  7.6122449   7.79591837  7.97959184  8.16326531  8.34693878  8.53061224
  8.71428571  8.89795918  9.08163265  9.26530612  9.44897959  9.63265306
  9.81632653 10.        ]
[ 1.   2.5  4.   5.5  7.   8.5 10. ]
[1.         2.28571429 3.57142857 4.85714286 6.14285714 7.42857143
 8.71428571]


Initalizing an Arrays

In [6]:
import numpy as np
# Create arrays of specified shapes
ones_array = np.ones((2, 3)) # Shape: (2, 3)
zeros_array = np.zeros((3, 2)) # Shape: (3, 2)
empty_array = np.empty((2, 2)) # Shape: (2, 2)
identity_matrix = np.eye(3) # Shape: (3, 3)
print(ones_array, zeros_array, empty_array, identity_matrix, sep='\n\n')


[[1. 1. 1.]
 [1. 1. 1.]]

[[0. 0.]
 [0. 0.]
 [0. 0.]]

[[6.55028425e-310 2.17643901e-316]
 [2.23322415e-316 6.55028425e-310]]

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


Using ndarray

In [7]:
import numpy as np
array_from_list = np.array([1, 2, 3]) # [1 2 3]
array_from_tuple = np.array((4, 5, 6)) # [4 5 6]
array_from_nested_list = np.array([[1, 2, 3], [4, 5, 6]]) # [[1 2 3] [4 5 6]]
print(array_from_list, array_from_tuple, array_from_nested_list, sep='\n')

[1 2 3]
[4 5 6]
[[1 2 3]
 [4 5 6]]


Using shape of an existing array

In [8]:
import numpy as np
# Existing Array of Shape:(2,3)
arr = np.array([[1, 2, 3], [4, 5, 6]])
# Creating Array with Shape of existing array:
zeros, ones, empty = np.zeros(arr.shape), np.ones(arr.shape), np.empty(arr.shape) # Shape: (2,3)
print(arr, zeros, ones, empty, sep='\n')


[[1 2 3]
 [4 5 6]]
[[0. 0. 0.]
 [0. 0. 0.]]
[[1. 1. 1.]
 [1. 1. 1.]]
[[2.13134030e-316 0.00000000e+000 0.00000000e+000]
 [8.95593996e-312 2.10966031e-321 4.23754071e-307]]


Implementing np.hstack() and np.vstack()

In [9]:
import numpy as np
arr1 = np.array([[1, 2], [3, 4]]) # Shape: (2, 2)
arr2 = np.array([[5, 6], [7, 8]]) # Shape: (2, 2)
# Stacking examples
hstacked = np.hstack((arr1, arr2))
# Output hstack: [[1 2 5 6] [3 4 7 8]]
vstacked = np.vstack((arr1, arr2))
# Output vstack[[1 2] [3 4] [5 6] [7 8]]
colstacked = np.column_stack((arr1[0], arr2[0]))
# Output column stack: [[1 5] [2 6]]
print(hstacked, vstacked, colstacked, sep='\n')


[[1 2 5 6]
 [3 4 7 8]]
[[1 2]
 [3 4]
 [5 6]
 [7 8]]
[[1 5]
 [2 6]]


Implementing np.concatenate()

In [10]:
import numpy as np

# Creating two 3D arrays
arr1 = np.array([[[1], [2]], [[3], [4]]])  # Shape: (2, 2, 1)
arr2 = np.array([[[5], [6]], [[7], [8]]])  # Shape: (2, 2, 1)

# Concatenate along axis 0 (vertical stacking)
concat_axis0 = np.concatenate((arr1, arr2), axis=0)
# Shape of concat_axis0: (4, 2, 1), arr1 is stacked on top of arr2

# Concatenate along axis 1 (horizontal stacking)
concat_axis1 = np.concatenate((arr1, arr2), axis=1)
# Shape of concat_axis1: (2, 4, 1), arr1 and arr2 are placed side by side

# Concatenate along axis 2 (depth stacking)
concat_axis2 = np.concatenate((arr1, arr2), axis=2)
# Shape of concat_axis2: (2, 2, 2), arr1 and arr2 are stacked in depth (third dimension)

# Displaying the results
print("Concatenation along axis 0 (Vertical stacking):\n", concat_axis0)
print("Shape of concatenated array along axis 0:", concat_axis0.shape)

print("\nConcatenation along axis 1 (Horizontal stacking):\n", concat_axis1)
print("Shape of concatenated array along axis 1:", concat_axis1.shape)

print("\nConcatenation along axis 2 (Depth stacking):\n", concat_axis2)
print("Shape of concatenated array along axis 2:", concat_axis2.shape)

Concatenation along axis 0 (Vertical stacking):
 [[[1]
  [2]]

 [[3]
  [4]]

 [[5]
  [6]]

 [[7]
  [8]]]
Shape of concatenated array along axis 0: (4, 2, 1)

Concatenation along axis 1 (Horizontal stacking):
 [[[1]
  [2]
  [5]
  [6]]

 [[3]
  [4]
  [7]
  [8]]]
Shape of concatenated array along axis 1: (2, 4, 1)

Concatenation along axis 2 (Depth stacking):
 [[[1 5]
  [2 6]]

 [[3 7]
  [4 8]]]
Shape of concatenated array along axis 2: (2, 2, 2)


## 7.1 Visualizing Array Combination Operations

Dimension Mismatch Sample Solution

In [11]:
import numpy as np

# Incorrect shapes
arr1 = np.array([[1, 2], [3, 4]])  # shape (2, 2)
arr2 = np.array([5, 6])            # shape (2,)

# Attempting horizontal stack
try:
    hstacked = np.hstack((arr1, arr2))  # Error due to dimension mismatch
except ValueError as e:
    print(f"Error: {e}")

# Solution: Reshape arr2 to match arr1’s dimensions
arr2 = np.array([5, 6]).reshape(2, 1)  # shape (2, 1)
hstacked = np.hstack((arr1, arr2))      # No error
print(hstacked)

Error: all the input arrays must have same number of dimensions, but the array at index 0 has 2 dimension(s) and the array at index 1 has 1 dimension(s)
[[1 2 5]
 [3 4 6]]


2. Shape Misalignment for Stacking:

In [12]:
import numpy as np
arr1 = np.array([[1, 2, 3], [4, 5, 6]]) # shape (2, 3)
arr2 = np.array([[7, 8], [9, 10]]) # shape (2, 2)
# Attempting horizontal stack (this will fail)
try:
   hstacked = np.hstack((arr1, arr2)) # Error due to shape mismatch
except ValueError as e:
 print(f"Error: {e}")
# Solution: Make sure the number of rows in both arrays is the same before stacking arr2 = np.array([[7, 8, 9], [10, 11, 12]]) # shape (2, 3)
hstacked = np.hstack((arr1, arr2)) # Works fine
print(hstacked)


[[ 1  2  3  7  8]
 [ 4  5  6  9 10]]


3. Inconsistent Data Types:

In [13]:
arr1 = np.array([[1, 2], [3, 4]], dtype=int)
arr2 = np.array([5, 6], dtype=float)
# Attempting horizontal stack (this will work, but can lead to unexpected results) hstacked = np.hstack((arr1, arr2))
print(hstacked)
# Solution: Convert the data types to match each other explicitly, if needed:


[[ 1  2  3  7  8]
 [ 4  5  6  9 10]]


4. Too Many Dimensions in the Arrays

In [14]:
import numpy as np

# Define arrays
arr1 = np.array([[[1, 2], [3, 4]]])  # shape (1, 2, 2)
arr2 = np.array([[5, 6], [7, 8]])     # shape (2, 2)

# Attempting stacking with 3D and 2D arrays
try:
    vstacked = np.vstack((arr1, arr2))  # Error due to too many dimensions
except ValueError as e:
    print(f"Error: {e}")

# Solution: Reshape arr1 to be 2D (or flatten it) before stacking
arr1 = arr1.reshape(2, 2)  # Now it’s shape (2, 2)
vstacked = np.vstack((arr1, arr2))  # Works fine
print(vstacked)

Error: all the input arrays must have same number of dimensions, but the array at index 0 has 3 dimension(s) and the array at index 1 has 2 dimension(s)
[[1 2]
 [3 4]
 [5 6]
 [7 8]]


## 8 Array Slicing and Indexing.

Array Indexing and Slicing

In [15]:
import numpy as np
# Creating a 2D array
array = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
# Accessing elements using indexing
print(array[0, 1]) # Output: 2 (first row, second column)
print(array[-1, -1]) # Output: 9 (last row, last column)
# Slicing the array
print(array[0:2, 1:3]) # Output: [[2, 3], [5, 6]]


2
9
[[2 3]
 [5 6]]


8.1 Index Out of Bounds Error

In [16]:
import numpy as np
# Creating a 2D array
array = np.array([[1, 2, 3], [4, 5, 6]])
# Attempting to access an element out of bounds
try:
 print(array[2, 1])
except IndexError as e:
 print(f"Error: {e}")


Error: index 2 is out of bounds for axis 0 with size 2


## 9 Array Mathematics with NumPy.

9.1 Universal Functions

In [17]:
import numpy as np
# Example of basic ufuncs: square and square root
arr = np.array([1, 4, 9, 16])
# Square every element
squared = np.square(arr)
# Square root of every element
sqrt = np.sqrt(arr)
print(f"Squared: {squared}")
print(f"Square root: {sqrt}")

Squared: [  1  16  81 256]
Square root: [1. 2. 3. 4.]


Example - Mathematical ufuncs

In [18]:
import numpy as np

# Trigonometric ufuncs
angles = np.array([0, np.pi / 4, np.pi / 2])  # Define an array of angles in radians
sine_values = np.sin(angles)  # Compute the sine of each angle
print("Sine values:", sine_values)

# Logarithmic ufuncs
arr = np.array([1, 10, 100])  # Define an array of positive numbers
log_values = np.log(arr)  # Compute the natural logarithm of each element
print("Logarithmic values:", log_values)

Sine values: [0.         0.70710678 1.        ]
Logarithmic values: [0.         2.30258509 4.60517019]


9.2 Linear Algebra with np.linalg.

In [19]:
import numpy as np
# Matrix multiplication
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])
product = np.dot(A, B)
# Compute determinant
det = np.linalg.det(A)
# Compute inverse
inverse = np.linalg.inv(A)
# Compute eigenvalues and eigenvectors
eigvals, eigvecs = np.linalg.eig(A)
print(f"Matrix Product:\n{product}")
print(f"Determinant: {det}")
print(f"Inverse:\n{inverse}")
print(f"Eigenvalues: {eigvals}, Eigenvectors:\n{eigvecs}")

Matrix Product:
[[19 22]
 [43 50]]
Determinant: -2.0000000000000004
Inverse:
[[-2.   1. ]
 [ 1.5 -0.5]]
Eigenvalues: [-0.37228132  5.37228132], Eigenvectors:
[[-0.82456484 -0.41597356]
 [ 0.56576746 -0.90937671]]


9.3 Broadcasting

In [20]:
import numpy as np
A = np.array([[1], [2], [3]]) # Shape (3,1) 3 rows, 1 column
B = np.array([10, 20, 30]) # Shape (3,) 1D array (3 elements)
C = A + B # Works! Result shape: (3,3)
print(C)

[[11 21 31]
 [12 22 32]
 [13 23 33]]


In [21]:
A = np.array([[1, 2, 3]]) # Shape (1,3)
B = np.array([[4], [5], [6]]) # Shape (3,1)
C = A + B #Works! Result shape: (3,3)
print(C)


[[5 6 7]
 [6 7 8]
 [7 8 9]]


In [22]:
A = np.ones((3,2)) # Shape (3,2)
B = np.ones((2,3)) # Shape (2,3)
C = A + B #ValueError: operands could not be broadcast together


ValueError: operands could not be broadcast together with shapes (3,2) (2,3) 

9.4 Vectorization

In [23]:
import numpy as np
import time

# Create two large arrays
size = 10**6
a = np.random.rand(size)  # Array of random numbers
b = np.random.rand(size)  # Array of random numbers

# Using a loop
start_time = time.time()
result_loop = np.zeros(size)  # Initialize an array to store the result
for i in range(size):
    result_loop[i] = a[i] + b[i]  # Element-wise addition in a loop
end_time = time.time()
loop_time = end_time - start_time
print(f"Loop Execution Time: {loop_time:.5f} seconds")

# Using NumPy’s vectorized operations
start_time = time.time()
result_vectorized = a + b  # Element-wise addition using NumPy's vectorized operations
end_time = time.time()
vectorized_time = end_time - start_time
print(f"Vectorized Execution Time: {vectorized_time:.5f} seconds")

# Performance improvement factor
speedup = loop_time / vectorized_time
print(f"Vectorization is approximately {speedup:.2f} times faster!")

Loop Execution Time: 0.47527 seconds
Vectorized Execution Time: 0.00608 seconds
Vectorization is approximately 78.13 times faster!



# 10 To - Do - NumPy

10.1 Basic Vector and Matrix Operation with Numpy.

Problem - 1: Array Creation:
Complete the following Tasks:
1. Initialize an empty array with size 2X2.
2. Initialize an all one array with size 4X2.
3. Return a new array of given shape and type, filled with fill value.{Hint: np.full} 4. Return a new array of zeros with same shape and type as a given array.{Hint: np.zeros like} 5. Return a new array of ones with same shape and type as a given array.{Hint: np.ones like} 6. For an existing list new_list = [1,2,3,4] convert to an numpy array.{Hint: np.array()}

In [24]:
import numpy as np
empty_array = np.empty((2, 2))
print("1. Empty array of size 2x2:\n", empty_array)
ones_array = np.ones((4, 2))
print("\n2. All ones array of size 4x2:\n", ones_array)

fill_value_array = np.full((3, 3), 5)
print("\n3. Array filled with value 5:\n", fill_value_array)
given_array = np.array([[1, 2], [3, 4]])
zeros_like_array = np.zeros_like(given_array)
print("\n4. Array of zeros with same shape as given array:\n", zeros_like_array)

ones_like_array = np.ones_like(given_array)
print("\n5. Array of ones with same shape as given array:\n", ones_like_array)

new_list = [1, 2, 3, 4]
numpy_array = np.array(new_list)
print("\n6. Numpy array from list:\n", numpy_array)

1. Empty array of size 2x2:
 [[2.5e-323 3.0e-323]
 [3.5e-323 4.0e-323]]

2. All ones array of size 4x2:
 [[1. 1.]
 [1. 1.]
 [1. 1.]
 [1. 1.]]

3. Array filled with value 5:
 [[5 5 5]
 [5 5 5]
 [5 5 5]]

4. Array of zeros with same shape as given array:
 [[0 0]
 [0 0]]

5. Array of ones with same shape as given array:
 [[1 1]
 [1 1]]

6. Numpy array from list:
 [1 2 3 4]


Problem - 2: Array Manipulation: Numerical Ranges and Array indexing: Complete the following tasks:
1. Create an array with values ranging from 10 to 49. {Hint:np.arrange()}.
2. Create a 3X3 matrix with values ranging from 0 to 8.
{Hint:look for np.reshape()}
3. Create a 3X3 identity matrix.{Hint:np.eye()}
4. Create a random array of size 30 and find the mean of the array.
{Hint:check for np.random.random() and array.mean() function}
5. Create a 10X10 array with random values and find the minimum and maximum values. 6. Create a zero array of size 10 and replace 5th element with 1.
7. Reverse an array arr = [1,2,0,0,4,0].
8. Create a 2d array with 1 on border and 0 inside.
9. Create a 8X8 matrix and fill it with a checkerboard pattern.

In [25]:
import numpy as np
arr1 = np.arange(10, 50)
print("1. Array with values from 10 to 49:\n", arr1)

arr2 = np.arange(9).reshape(3, 3)
print("\n2. 3x3 matrix with values from 0 to 8:\n", arr2)

identity_matrix = np.eye(3)
print("\n3. 3x3 identity matrix:\n", identity_matrix)
random_array = np.random.random(30)
mean_value = random_array.mean()
print("\n4. Random array of size 30:\n", random_array)
print("Mean of the array:", mean_value)

random_matrix = np.random.random((10, 10))
min_value = random_matrix.min()
max_value = random_matrix.max()
print("\n5. 10x10 random matrix:\n", random_matrix)
print("Minimum value:", min_value)
print("Maximum value:", max_value)

zero_array = np.zeros(10)
zero_array[4] = 1
print("\n6. Zero array with 5th element replaced by 1:\n", zero_array)

arr = np.array([1, 2, 0, 0, 4, 0])
reversed_arr = arr[::-1]
print("\n7. Reversed array:\n", reversed_arr)
border_array = np.ones((5, 5))
border_array[1:-1, 1:-1] = 0
print("\n8. 2D array with 1 on the border and 0 inside:\n", border_array)

checkerboard = np.zeros((8, 8), dtype=int)
checkerboard[1::2, ::2] = 1
checkerboard[::2, 1::2] = 1
print("\n9. 8x8 checkerboard pattern:\n", checkerboard)

1. Array with values from 10 to 49:
 [10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33
 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]

2. 3x3 matrix with values from 0 to 8:
 [[0 1 2]
 [3 4 5]
 [6 7 8]]

3. 3x3 identity matrix:
 [[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]

4. Random array of size 30:
 [0.15040284 0.53768191 0.03614303 0.77515916 0.42765516 0.09974697
 0.17301288 0.46763597 0.57584854 0.18422586 0.52573853 0.30795538
 0.53446549 0.15064783 0.01942089 0.47785727 0.56024331 0.74404387
 0.48223816 0.19156697 0.87741396 0.65396972 0.75129061 0.59368361
 0.21788998 0.72934756 0.15866095 0.29807832 0.99664591 0.18793433]
Mean of the array: 0.4295534986036969

5. 10x10 random matrix:
 [[0.29220812 0.92423813 0.30423575 0.30241086 0.25296959 0.08979985
  0.48930313 0.48430679 0.72318279 0.35573502]
 [0.3669464  0.46641278 0.32745832 0.56098439 0.08403867 0.37977705
  0.14528582 0.1272432  0.65845584 0.90780851]
 [0.86343287 0.98447401 0.72724437 0.07447985 0.

Problem - 3: Array Operations:
For the following arrays:
x = np.array([[1,2],[3,5]]) and y = np.array([[5,6],[7,8]]);
v = np.array([9,10]) and w = np.array([11,12]);
Complete all the task using numpy:
1. Add the two array.
2. Subtract the two array.
3. Multiply the array with any integers of your choice.
4. Find the square of each element of the array.
5. Find the dot product between: v(and)w ; x(and)v ; x(and)y.
6. Concatenate x(and)y along row and Concatenate v(and)w along column.
{Hint:try np.concatenate() or np.vstack() functions.
7. Concatenate x(and)v; if you get an error, observe and explain why did you get the error?


In [26]:
import numpy as np
x = np.array([[1, 2], [3, 5]])
y = np.array([[5, 6], [7, 8]])
v = np.array([9, 10])
w = np.array([11, 12])

add_xy = x + y
print("Addition of x and y:\n", add_xy)
subtract_xy = x - y
print("Subtraction of x and y:\n", subtract_xy)
multiply_x = x * 3
print("Multiplication of x with 3:\n", multiply_x)
square_x = np.square(x)
print("Square of each element in x:\n", square_x)
dot_vw = np.dot(v, w)
print("Dot product of v and w:", dot_vw)
dot_xv = np.dot(x, v)
print("Dot product of x and v:\n", dot_xv)
dot_xy = np.dot(x, y)
print("Dot product of x and y:\n", dot_xy)
concat_xy = np.concatenate((x, y), axis=0)
print("Concatenation of x and y along rows:\n", concat_xy)
concat_vw = np.column_stack((v, w))
print("Concatenation of v and w along columns:\n", concat_vw)
try:
    concat_xv = np.concatenate((x, v), axis=0)
    print("Concatenation of x and v:\n", concat_xv)
except ValueError as e:
    print("Error while concatenating x and v:", e)


Addition of x and y:
 [[ 6  8]
 [10 13]]
Subtraction of x and y:
 [[-4 -4]
 [-4 -3]]
Multiplication of x with 3:
 [[ 3  6]
 [ 9 15]]
Square of each element in x:
 [[ 1  4]
 [ 9 25]]
Dot product of v and w: 219
Dot product of x and v:
 [29 77]
Dot product of x and y:
 [[19 22]
 [50 58]]
Concatenation of x and y along rows:
 [[1 2]
 [3 5]
 [5 6]
 [7 8]]
Concatenation of v and w along columns:
 [[ 9 11]
 [10 12]]
Error while concatenating x and v: all the input arrays must have same number of dimensions, but the array at index 0 has 2 dimension(s) and the array at index 1 has 1 dimension(s)


Problem - 4: Matrix Operations:
• For the following arrays:
A = np.array([[3,4],[7,8]]) and B = np.array([[5,3],[2,1]]);
Prove following with Numpy:
1. Prove A.A−1 = I.
2. Prove AB ̸= BA.
3. Prove (AB)T = BTAT.
• Solve the following system of Linear equation using Inverse Methods.
2x − 3y + z = −1
x − y + 2z = −3
3x + y − z = 9
{Hint: First use Numpy array to represent the equation in Matrix form. Then Solve for: AX = B}
• Now: solve the above equation using np.linalg.inv function.{Explore more about ”linalg” func tion of Numpy}


In [27]:
import numpy as np
A = np.array([[3, 4], [7, 8]])
B = np.array([[5, 3], [2, 1]])

A_inv = np.linalg.inv(A)
I = np.dot(A, A_inv)

print("A * A^(-1) (Should be Identity Matrix):\n", np.round(I, decimals=5))
AB = np.dot(A, B)
BA = np.dot(B, A)

print("AB:\n", AB)
print("BA:\n", BA)
print("Is AB equal to BA? ", np.array_equal(AB, BA))

AB_T = np.transpose(AB)
BT_AT = np.dot(np.transpose(B), np.transpose(A))

print("(AB)^T:\n", AB_T)
print("B^T * A^T:\n", BT_AT)
print("Is (AB)^T equal to B^T * A^T? ", np.array_equal(AB_T, BT_AT))
# Given system of equations:
# 2x - 3y +  z  = -1
#  x -  y + 2z  = -3
# 3x +  y -  z  =  9
A_matrix = np.array([[2, -3, 1], [1, -1, 2], [3, 1, -1]])
B_matrix = np.array([-1, -3, 9])
A_inv_matrix = np.linalg.inv(A_matrix)
X = np.dot(A_inv_matrix, B_matrix)
print("Solution for [x, y, z]:", X)


A * A^(-1) (Should be Identity Matrix):
 [[1. 0.]
 [0. 1.]]
AB:
 [[23 13]
 [51 29]]
BA:
 [[36 44]
 [13 16]]
Is AB equal to BA?  False
(AB)^T:
 [[23 51]
 [13 29]]
B^T * A^T:
 [[23 51]
 [13 29]]
Is (AB)^T equal to B^T * A^T?  True
Solution for [x, y, z]: [ 2.  1. -2.]


### 10.2 Experiment: How Fast is Numpy?

1. Element-wise Addition:
• Using Python Lists, perform element-wise addition of two lists of size 1, 000, 000. Measure and Print the time taken for this operation.
• Using Numpy Arrays, Repeat the calculation and measure and print the time taken for this operation.


In [28]:
import numpy as np
import time

size = 1_000_000
list1 = list(range(size))
list2 = list(range(size))
start_time = time.time()
result_list = [a + b for a, b in zip(list1, list2)]
end_time = time.time()
print(f"Time taken using Python lists: {end_time - start_time:.5f} seconds")
array1 = np.arange(size)
array2 = np.arange(size)
start_time = time.time()
result_array = array1 + array2
end_time = time.time()
print(f"Time taken using NumPy arrays: {end_time - start_time:.5f} seconds")


Time taken using Python lists: 0.07035 seconds
Time taken using NumPy arrays: 0.00499 seconds


2. Element-wise Multiplication
• Using Python Lists, perform element-wise multiplication of two lists of size 1, 000, 000. Measure and Print the time taken for this operation.
• Using Numpy Arrays, Repeat the calculation and measure and print the time taken for this operation.


In [29]:
import numpy as np
import time

size = 1_000_000
list1 = list(range(1, size + 1))
list2 = list(range(1, size + 1))

start_time = time.time()
result_list = [a * b for a, b in zip(list1, list2)]
end_time = time.time()
print(f"Time taken using Python lists: {end_time - start_time:.5f} seconds")
array1 = np.arange(1, size + 1)
array2 = np.arange(1, size + 1)

start_time = time.time()
result_array = array1 * array2
end_time = time.time()
print(f"Time taken using NumPy arrays: {end_time - start_time:.5f} seconds")


Time taken using Python lists: 0.16465 seconds
Time taken using NumPy arrays: 0.00497 seconds


3. Dot Product
• Using Python Lists, compute the dot product of two lists of size 1, 000, 000. Measure and Print the time taken for this operation.
• Using Numpy Arrays, Repeat the calculation and measure and print the time taken for this operation.


In [30]:
import numpy as np
import time
size = 1_000_000
list1 = list(range(1, size + 1))
list2 = list(range(1, size + 1))

start_time = time.time()
dot_product_list = sum(a * b for a, b in zip(list1, list2))
end_time = time.time()
print(f"Dot product using Python lists: {dot_product_list}")
print(f"Time taken using Python lists: {end_time - start_time:.5f} seconds")

array1 = np.arange(1, size + 1)
array2 = np.arange(1, size + 1)
start_time = time.time()
dot_product_array = np.dot(array1, array2)
end_time = time.time()
print(f"Dot product using NumPy arrays: {dot_product_array}")
print(f"Time taken using NumPy arrays: {end_time - start_time:.5f} seconds")


Dot product using Python lists: 333333833333500000
Time taken using Python lists: 0.08774 seconds
Dot product using NumPy arrays: 333333833333500000
Time taken using NumPy arrays: 0.00249 seconds


4. Matrix Multiplication
• Using Python lists, perform matrix multiplication of two matrices of size 1000x1000. Mea sure and print the time taken for this operation.
• Using NumPy arrays, perform matrix multiplication of two matrices of size 1000x1000. Measure and print the time taken for this operation.


In [31]:
import numpy as np
import time


size = 1000
matrix1 = [[i * j for j in range(size)] for i in range(size)]
matrix2 = [[j * i for j in range(size)] for i in range(size)]
start_time = time.time()
result_matrix = [[sum(a * b for a, b in zip(row, col)) for col in zip(*matrix2)] for row in matrix1]
end_time = time.time()
print(f"Time taken using Python lists: {end_time - start_time:.5f} seconds")
array1 = np.random.rand(size, size)
array2 = np.random.rand(size, size)
start_time = time.time()
result_array = np.dot(array1, array2)
end_time = time.time()
print(f"Time taken using NumPy arrays: {end_time - start_time:.5f} seconds")


Time taken using Python lists: 224.56328 seconds
Time taken using NumPy arrays: 0.09899 seconds
